### PR TITLE
Specialize `_matmul!` on `ForwardDiff.Dual`s

### LINKED ISSUE #116: Combine with ForwardDiff: `ERROR: InexactError: trunc(Int64, NaN)`
Currently, Octavian.jl does not seem to support AD using ForwardDiff.jl:
```julia
julia> using Pkg; Pkg.activate(temp=true); Pkg.add(["ForwardDiff", "Octavian"])
[...]
  [f6369f11] + ForwardDiff v0.10.21
  [6fd5a793] + Octavian v0.3.4
[...]

julia> using ForwardDiff, Octavian, LinearAlgebra

julia> u = rand(50, 50); du = similar(u); A = rand(size(u, 1), size(u, 1));

julia> config = ForwardDiff.JacobianConfig(nothing, du, u);

julia> ForwardDiff.jacobian((du, u) -> mul!(du, A, u), du, u, config) ≈ kron(I(size(u, 2)), A) # stdlib is fine
true

julia> ForwardDiff.jacobian((du, u) -> matmul!(du, A, u), du, u, config) ≈ kron(I(size(u, 2)), A)
ERROR: InexactError: trunc(Int64, NaN)
Stacktrace:
  [1] trunc
    @ ./float.jl:805 [inlined]
  [2] floor(#unused#::Type{Int64}, x::Float64)
    @ Base ./float.jl:367
  [3] #s3#19
    @ ~/.julia/packages/Static/A5kpr/src/float.jl:106 [inlined]
  [4] var"#s3#19"(M::Any, ::Any, #unused#::Any)
    @ Static ./none:0
  [5] (::Core.GeneratedFunctionStub)(::Any, ::Vararg{Any})
    @ Core ./boot.jl:580
  [6] block_sizes(#unused#::Val{ForwardDiff.Dual{Nothing, Float64, 12}}, W::StaticInt{0}, α::Static.StaticFloat64{0.0}, β::Static.StaticFloat64{0.0}, L₁ₑ::Static.StaticFloat64{1353.3202323303008}, L₂ₑ::Static.StaticFloat64{66157.2550994403})
    @ Octavian ~/.julia/packages/Octavian/LB784/src/block_sizes.jl:16
  [7] block_sizes
    @ ~/.julia/packages/Octavian/LB784/src/block_sizes.jl:10 [inlined]
  [8] block_sizes
    @ ~/.julia/packages/Octavian/LB784/src/block_sizes.jl:23 [inlined]
  [9] __matmul!(C::LayoutPointers.StridedPointer{ForwardDiff.Dual{Nothing, Float64, 12}, 2, 1, 0, (1, 2), Tuple{StaticInt{104}, Int64}, Tuple{StaticInt{0}, StaticInt{0}}}, A::LayoutPointers.StridedPointer{Float64, 2, 1, 0, (1, 2), Tuple{StaticInt{8}, Int64}, Tuple{StaticInt{0}, StaticInt{0}}}, B::LayoutPointers.StridedPointer{ForwardDiff.Dual{Nothing, Float64, 12}, 2, 1, 0, (1, 2), Tuple{StaticInt{104}, Int64}, Tuple{StaticInt{0}, StaticInt{0}}}, α::StaticInt{1}, β::StaticInt{0}, M::Int64, K::Int64, N::Int64, nthread::Nothing)
    @ Octavian ~/.julia/packages/Octavian/LB784/src/matmul.jl:343
 [10] _matmul!
    @ ~/.julia/packages/Octavian/LB784/src/matmul.jl:289 [inlined]
 [11] matmul!
    @ ~/.julia/packages/Octavian/LB784/src/matmul.jl:256 [inlined]
 [12] matmul!(C::Matrix{ForwardDiff.Dual{Nothing, Float64, 12}}, A::Matrix{Float64}, B::Matrix{ForwardDiff.Dual{Nothing, Float64, 12}})
    @ Octavian ~/.julia/packages/Octavian/LB784/src/matmul.jl:236
 [13] (::var"#9#10")(du::Matrix{ForwardDiff.Dual{Nothing, Float64, 12}}, u::Matrix{ForwardDiff.Dual{Nothing, Float64, 12}})
    @ Main ./REPL[9]:1
 [14] chunk_mode_jacobian(f!::var"#9#10", y::Matrix{Float64}, x::Matrix{Float64}, cfg::ForwardDiff.JacobianConfig{Nothing, Float64, 12, Tuple{Matrix{ForwardDiff.Dual{Nothing, Float64, 12}}, Matrix{ForwardDiff.Dual{Nothing, Float64, 12}}}})
    @ ForwardDiff ~/.julia/packages/ForwardDiff/UDrkY/src/jacobian.jl:223
 [15] jacobian(f!::Function, y::Matrix{Float64}, x::Matrix{Float64}, cfg::ForwardDiff.JacobianConfig{Nothing, Float64, 12, Tuple{Matrix{ForwardDiff.Dual{Nothing, Float64, 12}}, Matrix{ForwardDiff.Dual{Nothing, Float64, 12}}}}, ::Val{true})
    @ ForwardDiff ~/.julia/packages/ForwardDiff/UDrkY/src/jacobian.jl:40
 [16] jacobian(f!::Function, y::Matrix{Float64}, x::Matrix{Float64}, cfg::ForwardDiff.JacobianConfig{Nothing, Float64, 12, Tuple{Matrix{ForwardDiff.Dual{Nothing, Float64, 12}}, Matrix{ForwardDiff.Dual{Nothing, Float64, 12}}}})
    @ ForwardDiff ~/.julia/packages/ForwardDiff/UDrkY/src/jacobian.jl:36
 [17] top-level scope
    @ REPL[9]:1
```
I'm using Julia version 1.7.0-rc1.

### PR DESCRIPTION
**Pull request by ranocha**
I implemented the "simple" versions of multiplications involving only one input with dual numbers since that's what I need right now. A specialization on two dual input arguments would be more involved and could be implemented later. With this PR, trying to call `matmul!` with two dual inputs will result in a method error, clarifying that this needs to be done.

This will probably not be the most efficient version since I just used `@tturbo` instead of more sophisticated techniques used in other parts of Octavian. However, it makes something work that threw an error before, so it's a first step to make it work. It could be sped up even further later.

Closes #116

### CODE REVIEW

File: src/forward_diff.jl [hunk 1]
```diff
@@ -0,0 +1,36 @@
...
+    _C
+end
+
+# multiplication of dual matrix by standard vector/matrix from the right
+@inline function _matmul!(_C::AbstractVecOrMat{DualT}, _A::AbstractMatrix{DualT}, B::AbstractVecOrMat,
+                          α=One(), β=Zero(), nthread::Nothing=nothing, MKN=nothing, contig_axis=nothing) where {DualT <: ForwardDiff.Dual}
+    A = real_rep(_A)
+    C = real_rep(_C)
+
+    @tturbo for n ∈ indices((C, B), (3, 2)), m ∈ indices((C, A), 2), l in indices((C, A), 1)
```
[the remark this answers is not in the document]
Comment (chriselrod) [reply src/forward_diff.jl]: Yes, that looks good to me.

File: Project.toml [hunk 2]
```diff
@@ -9,6 +9,7 @@ IfElse = "615f187c-cbe4-4ef1-ba3b-2fcf58d6d173"
 LoopVectorization = "bdcacae8-1622-11e9-2a5c-532679323890"
 ManualMemory = "d125e4d3-2237-4719-b19c-fa641b8a4667"
 PolyesterWeave = "1d0040c9-8b98-4ee7-8388-3f51789ca0ad"
+Requires = "ae029012-a4dd-5104-9daa-d747884805df"
```
Comment (chriselrod) [inline Project.toml:12]: Also needs a compat bounds.
This caused the test failures.

Comment (ranocha) [reply Project.toml:12]: Oh, right :facepalm: Sorry. Should be fixed in f0536a4

File: src/forward_diff.jl [hunk 3]
```diff
@@ -0,0 +1,45 @@
...
+            ArrayInterface.is_dense(_C), ArrayInterface.is_column_major(_C)))
+        # we can avoid the reshape and call the standard method
+        A = reinterpret(T, _A)
+        C = reinterpret(T, _C)
+        _matmul!(C, A, B, α, β, nthread, MKN)
+    else
+        # we cannot use the standard method directly
+        A = real_rep(_A)
+        C = real_rep(_C)
+
```
Comment (chriselrod) [inline src/forward_diff.jl:34]: These lines aren't covered. Mind adding a test using a `view` or `Adjoint`/`Transpose` so that the check fails?

Comment (ranocha) [reply src/forward_diff.jl:34]: Good point; done in 10479ce

**Pull request merged**